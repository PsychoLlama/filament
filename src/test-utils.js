import { graphql } from 'graphql';
import nock from 'nock';

import { createHueLoaders, huerl } from './context';
import * as resolvers from './resolvers';
import schema from './schema';

export const bridge = nock(huerl());

export const query = async ([request]) => {
  const response = await graphql(schema, request, resolvers, {
    hue: createHueLoaders(),
  });

  expect(response.errors).toBeUndefined();

  return response.data;
};

export const createLight = (fields = {}) => ({
  uniqueid: '75:45:12:38:00:fa:0c:26-1b',
  type: 'Extended color light',
  manufacturername: 'Philips',
  swversion: '5.23.1.13452',
  name: 'Light name',
  modelid: 'LCT001',

  state: {
    xy: [0.3144, 0.3301],
    reachable: true,
    colormode: 'xy',
    effect: 'none',
    alert: 'none',
    hue: 34076,
    sat: 251,
    bri: 254,
    on: true,
    ct: 153,
  },

  ...fields,
});

export const createGroup = (fields = {}) => ({
  name: 'Group name',
  class: 'Hallway',
  lights: [1, 2],
  recycle: true,
  type: 'Room',

  action: {
    alert: 'select',
    colormode: 'ct',
    effect: 'none',
    xy: [0.5, 0.5],
    hue: 10000,
    sat: 254,
    bri: 254,
    on: true,
    ct: 250,
  },

  /* eslint-disable camelcase */
  state: { all_on: false, any_on: true },
  /* eslint-enable camelcase */

  ...fields,
});

export const createScene = fields => ({
  owner: 'kn82nNLskEnks208Nla32LnLkndaekuDNKnsnks8',
  lastupdated: '2017-10-30T02:10:46',
  appdata: { version: 1 },
  lights: ['41', '42'],
  name: 'Fall Colors',
  recycle: false,
  locked: false,
  picture: '',
  version: 2,

  ...fields,
});

export const createScenes = fields => ({
  KseUksCEskA9Al2: createScene(),
  skeCskess872Ck4: createScene(),

  ...fields,
});

export const createSchedule = fields => ({
  created: '2017-10-31T13:45:56',
  localtime: 'W124/T07:45:00',
  description: 'MyRoutine',
  time: 'W124/T13:45:00',
  name: 'Morning Alarm',
  status: 'enabled',
  recycle: true,
  command: {
    address: '/api/t0k3n/groups/11/state',
    body: { on: true },
    method: 'PUT',
  },

  ...fields,
});

export const createSchedules = fields => ({
  1: createSchedule(),
  2: createSchedule(),

  ...fields,
});
