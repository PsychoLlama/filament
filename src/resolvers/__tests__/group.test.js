import { query, bridge, createGroup, createLight } from '../../test-utils';
import toHexColorCode, { colorTypes } from '../../utils/color';

jest.mock('../../../bridge.json');

describe('Group resolver', () => {
  let endpoint;

  afterEach(() => endpoint.done());

  it('returns a group', async () => {
    const group = createGroup();
    endpoint = bridge.get('/groups/5').reply(200, group);

    const result = await query`{
      group(id: 5) { name }
    }`;

    expect(result).toEqual({
      group: { name: group.name },
    });
  });

  it('returns the correct data', async () => {
    const group = createGroup();
    endpoint = bridge.get('/groups/10').reply(200, group);

    const result = await query`{
      group(id: 10) {
        name type class anyOn allOn id
      }
    }`;

    expect(result).toMatchSnapshot();
  });

  it('resolves every light beneath', async () => {
    const group = createGroup({ lights: [1] });
    const light = createLight();
    endpoint = bridge.get('/groups/15').reply(200, group);
    const lightEndpoint = bridge.get('/lights').reply(200, { 1: light });

    const result = await query`{
      group(id: 15) {
        lights { name id }
      }
    }`;

    expect(result.group).toEqual({
      lights: [{ name: light.name, id: '1' }],
    });

    lightEndpoint.done();
  });

  describe('color', () => {
    it('returns black if the group is powered off', async () => {
      const group = createGroup();
      group.state.any_on = false; // eslint-disable-line

      endpoint = bridge.get('/groups/25').reply(200, group);

      const result = await query`{
        group(id: 25) { color }
      }`;

      expect(result.group.color).toBe('#000000');
    });

    it('returns a hex code', async () => {
      const group = createGroup();
      group.state.colormode = colorTypes.HSB;
      group.state.hue = 64000;
      group.state.sat = 200;
      group.state.bri = 150;

      endpoint = bridge.get('/groups/25').reply(200, group);

      const result = await query`{
        group(id: 25) { color }
      }`;

      expect(result.group.color).toBe(toHexColorCode(group.action));
    });

    it('understands temperature colors', async () => {
      const group = createGroup();
      group.state.colormode = colorTypes.TEMP;
      group.state.bri = 100;
      group.state.ct = 215;

      endpoint = bridge.get('/groups/30').reply(200, group);

      const result = await query`{
        group(id: 30) { color }
      }`;

      expect(result.group.color).toBe(toHexColorCode(group.action));
    });

    it('understands XY colorspace', async () => {
      const group = createGroup();
      group.state.colormode = colorTypes.XY;
      group.state.xy = [0.234, 0.567];

      endpoint = bridge.get('/groups/35').reply(200, group);

      const result = await query`{
        group(id: 35) { color }
      }`;

      expect(result.group.color).toBe(toHexColorCode(group.action));
    });
  });
});
