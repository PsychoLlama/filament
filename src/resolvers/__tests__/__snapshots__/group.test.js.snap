// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Group resolver returns the correct data 1`] = `
Object {
  "group": Object {
    "allOn": false,
    "anyOn": true,
    "class": "Hallway",
    "id": "10",
    "name": "Group name",
    "type": "Room",
  },
}
`;
